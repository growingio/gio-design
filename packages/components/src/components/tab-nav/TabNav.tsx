import React, { useMemo, useState, useEffect, useRef } from 'react';
import classNames from 'classnames';
import toArray from 'rc-util/lib/Children/toArray';
import { isNil } from 'lodash';
import { TabNavProps, TabNavItemProps } from './interface';
import useRefs from '../../utils/hooks/useRefs';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';

const TabNav = (props: TabNavProps, ref?: React.RefObject<HTMLDivElement>) => {
  const {
    children,
    prefixCls: customizePrefixCls,
    type = 'block',
    size = 'large',
    onChange,
    onTabClick,
    activeKey,
    defaultActiveKey = '',
  } = props;

  const [localActiveKey, setLocalActiveKey] = useState<string | number>(activeKey || defaultActiveKey);
  const [inkStyle, setInkStyle] = useState<{ left?: number; width?: number }>({});
  const wrapperRefKey = useRef<symbol>(Symbol('tabNav'));
  const [setRef, getRef] = useRefs<HTMLDivElement>();

  const prefixCls = usePrefixCls('tabnav', customizePrefixCls);
  const classString = classNames(prefixCls, `${prefixCls}-${type}`, {
    [`${prefixCls}-lg`]: size === 'large',
    [`${prefixCls}-md`]: size === 'middle',
    [`${prefixCls}-sm`]: size === 'small',
    [`${prefixCls}-xs`]: size === 'xs' && type === 'block',
  });

  const [tabNavKeys, tabNavChildren] = useMemo(() => {
    const _tabNavKeys: (string | number)[] = [];
    const _tabNavChildren = toArray(children).map((node: React.ReactElement<TabNavItemProps>, index) => {
      if (React.isValidElement(node) && node.type === TabNav.Item) {
        const { className, disabled, onClick, ...rest } = node.props;
        const _key = isNil(node.key) ? index : node.key;
        _tabNavKeys.push(_key);
        return (
          <TabNav.Item
            className={classNames(className, `${prefixCls}-item`, {
              [`${prefixCls}-item-active`]: localActiveKey === _key,
              [`${prefixCls}-item-disabled`]: disabled,
            })}
            prefixCls={prefixCls}
            disabled={disabled}
            key={_key}
            ref={setRef(_key)}
            onClick={(e) => {
              if (!disabled) {
                onClick?.(e);
                onTabClick?.(_key);
                if (isNil(activeKey)) {
                  setLocalActiveKey(_key);
                }
                if (localActiveKey !== _key) {
                  onChange?.(_key);
                }
              }
            }}
            {...rest}
          />
        );
      }
      return null;
    });
    return [_tabNavKeys, _tabNavChildren];
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [children, localActiveKey, onChange, onTabClick]);

  useMemo(() => {
    if (!tabNavKeys.includes(localActiveKey)) {
      setLocalActiveKey(tabNavKeys[0]);
    }
  }, [localActiveKey, tabNavKeys]);

  useMemo(() => {
    if (!isNil(activeKey) && tabNavKeys.includes(activeKey)) {
      setLocalActiveKey(activeKey);
    }
  }, [activeKey, tabNavKeys]);

  useEffect(() => {
    if (!isNil(getRef(localActiveKey)?.current) && !isNil(getRef(wrapperRefKey.current)?.current)) {
      const { left, width } = getRef(localActiveKey)!.current!.getBoundingClientRect();
      const wrapperLeft = getRef(wrapperRefKey.current)!.current!.getBoundingClientRect().left;
      setInkStyle({ left: left - wrapperLeft, width });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [localActiveKey, tabNavChildren]);

  return (
    <div className={classString} ref={setRef(wrapperRefKey.current, ref)}>
      {tabNavChildren}
      <div className={`${prefixCls}-ink-bar`} style={{ ...inkStyle }} />
    </div>
  );
};

TabNav.Item = React.forwardRef(
  ({ prefixCls, children, ...rest }: TabNavItemProps, ref: React.RefObject<HTMLDivElement>) => (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <div ref={ref} {...rest}>
      <div className={`${prefixCls}-item-btn`}>{children}</div>
    </div>
  )
);

export default TabNav;
