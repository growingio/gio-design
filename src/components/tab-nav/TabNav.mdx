import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import TabNav from './TabNav';

# TabNav 标签导航

<Subtitle>标签导航是两个或更多按钮段的集合。标签导航 TabNav 与 标签页 Tabs 的区别：点击切换标签页时，页面内容随之切换；点击分段控件，切换内容由页面决定。</Subtitle>

## 参数说明
<ArgsTable of={TabNav} />

## 代码演示

### 块状标签导航
一共四种尺寸：L（40px）、M（36px）、S（30px）、XS（24px）。

[Example](/?path=/story/basic-components-avatar--block)

### 线状标签导航
一共三种尺寸：L（40px）、M（36px）、S（30px）。

[Example](/?path=/story/basic-components-avatar--line)
