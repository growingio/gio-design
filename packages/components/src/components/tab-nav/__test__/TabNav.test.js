import React from 'react';
import { mount, render, shallow } from 'enzyme';
import TabNav from '../index';

describe('Testing TabNav', () => {
  const getTabNav = () => (
    <TabNav>
      <TabNav.Item>111</TabNav.Item>
      <TabNav.Item>222</TabNav.Item>
      <TabNav.Item>333</TabNav.Item>
      <TabNav.Item disabled>444</TabNav.Item>
    </TabNav>
  );

  it('should be stable', () => {
    const wrapper = render(getTabNav());
    expect(wrapper).toMatchSnapshot();
  });

  it('should be mount, setProps, unmount with no error', () => {
    expect(() => {
      const wrapper = mount(getTabNav());
      wrapper.setProps({ type: 'line' });
      wrapper.setProps({ type: 'small' });
      wrapper.unmount();
    }).not.toThrow();
  });

  test('props type', () => {
    const wrapper = shallow(getTabNav());
    expect(wrapper.exists('.gio-tabnav-block')).toBe(true);
    wrapper.setProps({ type: 'line' });
    expect(wrapper.exists('.gio-tabnav-line')).toBe(true);
  });

  test('props size', () => {
    const wrapper = shallow(getTabNav());
    expect(wrapper.exists('.gio-tabnav-lg')).toBe(true);
    wrapper.setProps({ size: 'middle' });
    expect(wrapper.exists('.gio-tabnav-md')).toBe(true);
    wrapper.setProps({ size: 'small' });
    expect(wrapper.exists('.gio-tabnav-sm')).toBe(true);
  });

  test('prop defaultActiveKey', () => {
    const wrapper = mount(
      <TabNav defaultActiveKey={1}>
        <TabNav.Item>111</TabNav.Item>
        <TabNav.Item>222</TabNav.Item>
        <TabNav.Item>333</TabNav.Item>
        <TabNav.Item disabled>444</TabNav.Item>
      </TabNav>
    );
    expect(wrapper.find('.gio-tabnav').childAt(1).exists('.gio-tabnav-item-active')).toBe(true);
  });

  test('prop activeKey', () => {
    const wrapper = mount(getTabNav());
    wrapper.setProps({ activeKey: 1 });
    expect(wrapper.find('.gio-tabnav').childAt(1).exists('.gio-tabnav-item-active')).toBe(true);
    wrapper.setProps({ activeKey: 2 });
    expect(wrapper.find('.gio-tabnav').childAt(2).exists('.gio-tabnav-item-active')).toBe(true);
    wrapper.find('.gio-tabnav-item').at(1).simulate('click');
    expect(wrapper.find('.gio-tabnav-item').at(1).exists('.gio-tabnav-item-active')).toBe(false);
  });

  test('prop onChange and onTabClick', () => {
    const onChange = jest.fn();
    const onTabClick = jest.fn();
    const wrapper = mount(getTabNav());
    wrapper.setProps({ onChange, onTabClick });
    wrapper.find('.gio-tabnav-item').at(1).simulate('click');
    expect(onChange).not.toHaveBeenCalled();
    expect(onTabClick).toHaveBeenCalled();
    wrapper.find('.gio-tabnav-item').at(2).simulate('click');
    expect(onChange).toHaveBeenCalled();
  });

  it('should be render rightly', () => {
    const wrapper = mount(getTabNav());
    expect(wrapper.find('.gio-tabnav').children()).toHaveLength(5);
    expect(wrapper.find('.gio-tabnav').childAt(0).exists('.gio-tabnav-item-active')).toBe(true);
    expect(wrapper.find('.gio-tabnav').childAt(3).exists('.gio-tabnav-item-disabled')).toBe(true);
  });

  it('should be render content rightly', () => {
    const wrapper = mount(getTabNav());
    expect(wrapper.find('.gio-tabnav-item-active').at(0).text()).toBe('111');
    wrapper.find('.gio-tabnav').childAt(1).simulate('click');
    expect(wrapper.find('.gio-tabnav-item-active').at(0).text()).toBe('222');
  });

  it('only render TabNav.Item', () => {
    const wrapper = mount(
      <TabNav defaultActiveKey={1}>
        <TabNav.Item>111</TabNav.Item>
        <TabNav.Item>222</TabNav.Item>
        <TabNav.Item>333</TabNav.Item>
        <TabNav.Item disabled>444</TabNav.Item>
        <div className="norender-div">我不会被渲染</div>
      </TabNav>
    );
    expect(wrapper.exists('.norender-div')).toBe(false);
  });
});
