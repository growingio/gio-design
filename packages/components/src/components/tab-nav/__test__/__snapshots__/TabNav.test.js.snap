// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TabNav should be stable 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "gio-tabnav gio-tabnav-block gio-tabnav-lg",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "gio-tabnav-item gio-tabnav-item-active",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "gio-tabnav-item-btn",
            },
            "children": Array [
              Object {
                "data": "111",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "gio-tabnav-item",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-tabnav-item-btn",
              },
              "children": Array [
                Object {
                  "data": "222",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "gio-tabnav-item",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "gio-tabnav-item-btn",
                },
                "children": Array [
                  Object {
                    "data": "333",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "gio-tabnav-item gio-tabnav-item-disabled",
                "disabled": "",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "gio-tabnav-item-btn",
                  },
                  "children": Array [
                    Object {
                      "data": "444",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "gio-tabnav-ink-bar",
                },
                "children": Array [],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "disabled": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "disabled": undefined,
              },
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "gio-tabnav-item",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "gio-tabnav-item-btn",
            },
            "children": Array [
              Object {
                "data": "222",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "gio-tabnav-item",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-tabnav-item-btn",
              },
              "children": Array [
                Object {
                  "data": "333",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "gio-tabnav-item gio-tabnav-item-disabled",
              "disabled": "",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "gio-tabnav-item-btn",
                },
                "children": Array [
                  Object {
                    "data": "444",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "gio-tabnav-ink-bar",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "disabled": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "disabled": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-tabnav-item gio-tabnav-item-active",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-tabnav-item-btn",
              },
              "children": Array [
                Object {
                  "data": "111",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "gio-tabnav-item",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "gio-tabnav-item-btn",
            },
            "children": Array [
              Object {
                "data": "333",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "gio-tabnav-item gio-tabnav-item-disabled",
            "disabled": "",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-tabnav-item-btn",
              },
              "children": Array [
                Object {
                  "data": "444",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "gio-tabnav-ink-bar",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "disabled": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "disabled": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-tabnav-item",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-tabnav-item-btn",
              },
              "children": Array [
                Object {
                  "data": "222",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "gio-tabnav-item gio-tabnav-item-active",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "gio-tabnav-item-btn",
                },
                "children": Array [
                  Object {
                    "data": "111",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "gio-tabnav-item gio-tabnav-item-disabled",
          "disabled": "",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "gio-tabnav-item-btn",
            },
            "children": Array [
              Object {
                "data": "444",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "gio-tabnav-ink-bar",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-tabnav-item",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-tabnav-item-btn",
              },
              "children": Array [
                Object {
                  "data": "333",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "gio-tabnav-item",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "gio-tabnav-item-btn",
                },
                "children": Array [
                  Object {
                    "data": "222",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "gio-tabnav-item gio-tabnav-item-active",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "gio-tabnav-item-btn",
                  },
                  "children": Array [
                    Object {
                      "data": "111",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "disabled": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "disabled": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "gio-tabnav-ink-bar",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "gio-tabnav-item gio-tabnav-item-disabled",
            "disabled": "",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "gio-tabnav-item-btn",
              },
              "children": Array [
                Object {
                  "data": "444",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "gio-tabnav-item",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "gio-tabnav-item-btn",
                },
                "children": Array [
                  Object {
                    "data": "333",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "gio-tabnav-item",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "gio-tabnav-item-btn",
                  },
                  "children": Array [
                    Object {
                      "data": "222",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "gio-tabnav-item gio-tabnav-item-active",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "gio-tabnav-item-btn",
                    },
                    "children": Array [
                      Object {
                        "data": "111",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "disabled": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "disabled": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
